### PR TITLE
[controller_runtime] operator/registration: switch to Kubebuilder client

### LINKED ISSUE #239: Kubebuilder: Refactor registration server to use controller runtime
The [machine registration](https://github.com/rancher/elemental-operator/blob/main/pkg/server/register.go#L32) server is not using controller-runtime client. We should replace all occurrences of old API definitions and client-go client. This was partially implemented in the [POC](https://github.com/alexander-demicev/elemental-operator/blob/ctrlruntime/pkg/server/register.go)

### PR DESCRIPTION
**Pull request by fgiudici**
Fixes #239

### CODE REVIEW

File: pkg/server/register.go [hunk 1]
```diff
@@ -121,101 +119,135 @@ func upgrade(resp http.ResponseWriter, req *http.Request) (*websocket.Conn, erro
...
 
-func (i *InventoryServer) writeMachineInventoryCloudConfig(conn *websocket.Conn, inventory *elm.MachineInventory, registration *elm.MachineRegistration) error {
-	var err error
+func (i *InventoryServer) writeMachineInventoryCloudConfig(conn *websocket.Conn, inventory *elementalv1.MachineInventory, registration *elementalv1.MachineRegistration) error {
+	var sa *corev1.ServiceAccount
+
+	err := i.Get(i, types.NamespacedName{
+		Name:      registration.Status.ServiceAccountRef.Name,
+		Namespace: registration.Status.ServiceAccountRef.Namespace,
+	}, sa)
```
Comment (alexander-demicev) [inline pkg/server/register.go:228]:
```suggestion
	if err := i.Get(i, types.NamespacedName{
		Name:      registration.Status.ServiceAccountRef.Name,
		Namespace: registration.Status.ServiceAccountRef.Namespace,
	}, sa); err != nil {
```

**Review by mjura — APPROVED** [review]
LGTM

File: api/v1beta1/types.go [hunk 2]
```diff
@@ -20,67 +20,67 @@ import runtime "k8s.io/apimachinery/pkg/runtime"
...
 
 type Config struct {
 	// +optional
-	Elemental Elemental `json:"elemental,omitempty"`
+	Elemental Elemental `json:"elemental,omitempty" yaml:"elemental"`
 	// +kubebuilder:validation:Schemaless
 	// +kubebuilder:validation:XPreserveUnknownFields
 	// +optional
-	CloudConfig map[string]runtime.RawExtension `json:"cloud-config,omitempty"`
+	CloudConfig runtime.RawExtension `json:"cloud-config,omitempty" yaml:"cloud-config,omitempty"`
```
[the remark this answers is not in the document]
Comment (alexander-demicev) [reply api/v1beta1/types.go:85]: +1 to keeping `map[string]runtime.RawExtension` but won't save us from users having invalid values
This is still possible:
```yaml
cloud-config:
  example: "invalid-" \
value
```
Old code also has this issue, but it's generally a problem of unstructured API. Maybe we can add some validation in a webhook later.

[the remark this answers is not in the document]
Comment (davidcassany) [reply api/v1beta1/types.go:85]: Yes, keeping the map only guarantees we can access data by `config.CloudConfig[<somestring>]`, any further structure is not validated. Just forces `cloud-config` to include mapped elements, no literals neither arrays.

File: pkg/server/register.go (hunk 1 above)
Comment (fgiudici) [reply pkg/server/register.go:228]: The error is checked in the following line... but guess you that and you are hinting here that the right pattern for client runtime usage is the contract form `if err := i.Get(...`. Am I right?
Makes sense, will adhere to the format (if there is more just tell!).

File: api/v1beta1/types.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (fgiudici) [reply api/v1beta1/types.go:85]: I see your point (prevent things like `cloud-config: "my cloud-config"`), I didn't think about it.
Anyway the issue here is what emerges in the thread: we have no validation on cloud config data. Preventing the case above doesn't help much on validation IMO, invalid CloudConfigs are still possible in many ways.
We end up wrapping CloudConfig in JSON format to carry it around. This is what in the end is a RawExtension, a []byte field carrying a JSON representation.
Exchanging a single structured JSON with a map of JSONs from a single yaml... well, looks to me we are making it more messier. In the end we will have to convert the JSON to a single YAML file: having to construct it cycling through a map of JSONs looks uglier to me than having all in one JSON.
So, I still prefer the single JSON approach.
Now that I have ended my rant I can change it back to a RawExtension map 🤣  thanks for the discussion folks!

File: cmd/register/main.go [hunk 3]
```diff
@@ -173,8 +173,25 @@ func run(config config.Config) {
 		}
 		cloudInitURLs = append(cloudInitURLs, agentConfPath)
 
-		if len(config.CloudConfig) > 0 {
-			cloudInitPath, err := writeCloudInit(config.CloudConfig)
+		// runtime.RawExtension is basically a JSON container. We first marshal it to JSON,
+		// then unmarshall the JSON to a map[string]interface{} structure which is then converted
+		// to a yaml file.
```
[the remark this answers is not in the document]
Comment (fgiudici) [reply cmd/register/main.go:178]: runtime.RawExtension IS in json format. what the `MarshalJSON()` method do in our case is just to return RawExtension.Raw, which is a []byte.
I could have used RawExtension.Raw, but I think it is more clean to use the `MarshalJSON()` method to retrieve the content of the field. The error cannot happen in our case, but still better to check, one never knows how the runtime package may evolve 🤷🏼 
We than want to go directly from there (JSON) to the yaml file. Anyway, the quickest way is Unmarshal the JSON and then Marshal the data to YAML.
That said... I totally agree we should at least wrap this in our internal logic. This is part of the rework of the serialization that I would address in a separate commit on top of this PR.

**Comment by fgiudici** [comment]
> So far looks good to me, even I have concerns about the serialization and deserialization of data. We should find an approach that, al least, centralized all the ugliness in a single code portion, so it can be easily revisited in the future if needed.

Totally agree 👍🏼  Will work it on top of the current commits.

File: pkg/server/register.go (hunk 1 above)
Comment (alexander-demicev) [reply pkg/server/register.go:228]: inlining `if err := ...`  is just idiomatic in golang :D

**Review by alexander-demicev — APPROVED** [review]
thanks, all looks good 🎉

**Pull request merged**